### DISCUSSION TITLE
Parse TCP Payload containing OPC UA Data

### DISCUSSION THREAD

**Discussion by philipempl** [FreeOpcUa/python-opcua#1493] [Q&A] [answered]
Hi guys,

I use pyshark to sniff OPC UA network packets, which works quite well. When the OPC UA communication is based on extension objects, pyshark reaches its limits. I was thinking of parsing the TCP payloads natively to get a well structured packet. 

Is it possible to parse the hex payload and create OPC UA packages with this module?

If not, do you have any other idea how I can solve the problem?

Thank you very much and have a nice day.

With kind regards
Philip

**Comment by schroeder-** [comment, marked as answer]
The full code for decoding is in https://github.com/FreeOpcUa/python-opcua/blob/master/opcua/common/connection.py.
This is complex because there is security settings and chunking.

If you only want to decrypt ExtensionObjects you can use this function:

```
from opcua.ua.ua_binary import  extensionobject_from_binary
data = b'......'
buffer = ua.utils.Buffer(data)
ext_obj = extensionobject_from_binary(buffer)
```


Anyway you have to load the extension objects from the server first. See: client.load_type_definitions()
